[nnc] Removed const from all fields in IR.

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62337
* __->__ #62336

This PR was generated by removing `const` for all types of nodes in NNC IR, and fixing compilation errors that were the result of this change. 

This is the first step in making all NNC mutations in-place.

Differential Revision: [D30049829](https://our.internmc.facebook.com/intern/diff/D30049829)

navahgar committed:
[nnc] Removed const from all fields in IR.

[ghstack-poisoned]

huiguoo commented:
This is great! Thanks for making it happen @navahgar !! 

One question, are we going to gradually add `const` back to `ir_visitor` and its children, or eventually visit IRs without `const`?

navahgar commented:
> One question, are we going to gradually add `const` back to `ir_visitor` and its children, or eventually visit IRs without `const`?

I tried doing that and it resulted in lot more `const_cast`s which looked more ugly IMO. So, personally, I don't want to add `const` back in `IRVisitor`. But we can revisit that later if necessary, since that can be done independently.

huiguoo commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -12,14 +12,14 @@ namespace jit {
 namespace tensorexpr {
 
 // Returns true if the TE fuser supports this conv2d.
-bool conv2dIsSupportedJit(const Node* node);
```
Should we keep jit::Node const?

navahgar replied:
That is a good point. We should definitely keep all data structures not belonging to NNC const. Modified that.

huiguoo commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -12,14 +12,14 @@ namespace jit {
-bool conv2dIsSupportedJit(const Node* node);
+bool conv2dIsSupportedJit(Node* node);
 // Returns true if the TE fuser supports this matmul.
-bool matmulIsSupported(const Node* node);
```
Same as above

navahgar replied:
Done.

huiguoo commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -167,10 +167,10 @@ class LLVMCodeGenImpl : public IRVisitor {
   llvm::Type* Int8PtrTy_;
   llvm::Type* VoidTy_;
 
-  std::unordered_map<const Var*, int> varToArg_;
```
We can keep `Var` to be `const` in this case for better coding.

navahgar replied:
I wasn't sure about this. I wanted to keep this PR simple in that it removes const from all NNC IR. But, in this case, since it is just a key to a map and making it const did not affect any other code, I'm okay doing that. 

Updated.

huiguoo commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -167,10 +167,10 @@ class LLVMCodeGenImpl : public IRVisitor {
   llvm::Type* VoidTy_;
 
-  std::unordered_map<const Var*, int> varToArg_;
-  std::unordered_map<const Var*, llvm::Value*> varToVal_;
```
Same as above

navahgar replied:
Same as above. Done.

huiguoo commented:
LGTM! Guess we can gradually revisit the affected code and add back small 'const's.

navahgar committed:
Update on "[nnc] Removed const from all fields in IR."


This PR was generated by removing `const` for all types of nodes in NNC IR, and fixing compilation errors that were the result of this change. 

This is the first step in making all NNC mutations in-place. 

[ghstack-poisoned]

navahgar commented:
> LGTM! Guess we can gradually revisit the affected code and add back small 'const's.

Thanks for the review. Sure, we could definitely consider adding back const's in some places later.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/half_support.h:
```diff
@@ -13,7 +13,7 @@ namespace tensorexpr {
 class HalfChecker : public IRVisitor {
  public:
   HalfChecker(const std::vector<CodeGen::BufferArg>& args) {
-    for (const auto& BA : args) {
```
I think we can keep `const` in `const auto&` - it probably doesn't create any issues with constness.

navahgar replied:
Done.

navahgar committed:
Update on "[nnc] Removed const from all fields in IR."


This PR was generated by removing `const` for all types of nodes in NNC IR, and fixing compilation errors that were the result of this change. 

This is the first step in making all NNC mutations in-place. 

[ghstack-poisoned]

navahgar committed:
Update on "[nnc] Removed const from all fields in IR."


This PR was generated by removing `const` for all types of nodes in NNC IR, and fixing compilation errors that were the result of this change. 

This is the first step in making all NNC mutations in-place. 

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30049829).

navahgar committed:
Update on "[nnc] Removed const from all fields in IR."


This PR was generated by removing `const` for all types of nodes in NNC IR, and fixing compilation errors that were the result of this change. 

This is the first step in making all NNC mutations in-place.

Differential Revision: [D30049829](https://our.internmc.facebook.com/intern/diff/D30049829)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30049829).

navahgar committed:
Update on "[nnc] Removed const from all fields in IR."


This PR was generated by removing `const` for all types of nodes in NNC IR, and fixing compilation errors that were the result of this change. 

This is the first step in making all NNC mutations in-place.

Differential Revision: [D30049829](https://our.internmc.facebook.com/intern/diff/D30049829)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30049829).

navahgar committed:
Update on "[nnc] Removed const from all fields in IR."


This PR was generated by removing `const` for all types of nodes in NNC IR, and fixing compilation errors that were the result of this change. 

This is the first step in making all NNC mutations in-place.

Differential Revision: [D30049829](https://our.internmc.facebook.com/intern/diff/D30049829)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30049829).
